plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/FieldEmbedding.kt

```diff
@@ -45,23 +49,29 @@ interface FieldEmbedding {
             AccessPolicy.ALWAYS_READABLE, AccessPolicy.ALWAYS_WRITEABLE -> null
         }
 
-    fun accessInvariantsForPredicate(): TypeInvariantEmbedding? =
-        when (accessPolicy) {
-            AccessPolicy.ALWAYS_READABLE -> FieldAccessTypeInvariantEmbedding(this, PermExp.WildcardPerm())
-            AccessPolicy.ALWAYS_INHALE_EXHALE, AccessPolicy.ALWAYS_WRITEABLE -> null
-        }
+    fun accessInvariantsForPredicate(): TypeInvariantEmbedding? = null
 }
 
 class UserFieldEmbedding(
     override val name: ScopedKotlinName,
     override val type: TypeEmbedding,
     override val symbol: FirPropertySymbol,
+    override val isUnique: Boolean,
 ) : FieldEmbedding {
     override val viperType = Type.Ref
     override val accessPolicy: AccessPolicy = if (symbol.isVal) AccessPolicy.ALWAYS_READABLE else AccessPolicy.ALWAYS_INHALE_EXHALE
+    override val includeInUniquePredicate: Boolean
+        get() = isUnique || type.isPrimitive
```

Comment:
I don't think this is quite right.  A unique predicate should give one permission to read and modify the members.  That includes primitives, but it also includes references.  The only special treatment unique members get is that you also get access *their* members as well, via a nested unique predicate.  For non-unique members, you get their RO predicate.